types/decimal.go

```diff
@@ -0,0 +1,356 @@
+	if !ok {
+		return d, ErrUnknownRequest("bad string to integer conversion")
+	}
+	if neg {
+		combined = new(big.Int).Neg(combined)
+	}
+	return Dec{combined}, nil
+}
+
+//nolint
+func (d Dec) IsZero() bool      { return (d.Int).Sign() == 0 } // Is equal to zero
+func (d Dec) Equal(d2 Dec) bool { return (d.Int).Cmp(d2.Int) == 0 }
+func (d Dec) GT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == 1 }             // greater than
+func (d Dec) GTE(d2 Dec) bool   { return (d.Int).Cmp(d2.Int) >= 0 }             // greater than or equal
+func (d Dec) LT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == -1 }            // less than
+func (d Dec) LTE(d2 Dec) bool   { return (d.Int).Cmp(d2.Int) <= 0 }             // less than or equal
+func (d Dec) Neg() Dec          { return Dec{new(big.Int).Neg(d.Int)} }         // Is equal to zero
+func (d Dec) Add(d2 Dec) Dec    { return Dec{new(big.Int).Add(d.Int, d2.Int)} } // addition
+func (d Dec) Sub(d2 Dec) Dec    { return Dec{new(big.Int).Sub(d.Int, d2.Int)} } // subtraction
+
+// multiplication
+func (d Dec) Mul(d2 Dec) Dec {
+	mul := new(big.Int).Mul(d.Int, d2.Int)
+	chopped := BankerRoundChop(mul, Precision)
+
+	if chopped.BitLen() > 255 {
+		panic("Int overflow")
+	}
+	return Dec{chopped}
+}
+
+// quotient
+func (d Dec) Quo(d2 Dec) Dec {
+
+	// multiply precision twice
+	mul := new(big.Int).Mul(d.Int, precisionReuse)
+	mul.Mul(mul, precisionReuse)
+
+	quo := new(big.Int).Quo(mul, d2.Int)
+	chopped := BankerRoundChop(quo, Precision)
+	return Dec{chopped}
+}
+
+func (d Dec) String() string {
+	str := d.ToLeftPaddedWithDecimals(Precision)
+	placement := len(str) - Precision
+	if placement < 0 {
+		panic("too few decimal digits")
+	}
+	return str[:placement] + "." + str[placement:]
+}
+
+// TODO panic if negative or if totalDigits < len(initStr)???
+// evaluate as an integer and return left padded string
+func (d Dec) ToLeftPaddedWithDecimals(totalDigits int8) string {
+	intStr := d.Int.String()
+	fcode := `%0` + strconv.Itoa(int(totalDigits)) + `s`
+	return fmt.Sprintf(fcode, intStr)
+}
+
+// TODO panic if negative or if totalDigits < len(initStr)???
+// evaluate as an integer and return left padded string
+func (d Dec) ToLeftPadded(totalDigits int8) string {
+	chopped := BankerRoundChop(d.Int, Precision)
+	intStr := chopped.String()
+	fcode := `%0` + strconv.Itoa(int(totalDigits)) + `s`
+	return fmt.Sprintf(fcode, intStr)
+}
+
+//     ____
+//  __|    |__   "chop 'em
+//       ` \     round!"
+// ___||  ~  _     -bankers
+// |         |      __
+// |       | |   __|__|__
+// |_____:  /   | $$$    |
+//              |________|
+
+// nolint - go-cyclo
+// chop of n digits, and banker round the digits being chopped off
+// Examples:
+//   BankerRoundChop(1005, 1) = 100
+//   BankerRoundChop(1015, 1) = 102
+//   BankerRoundChop(1500, 3) = 2
+func BankerRoundChop(d *big.Int, n int64) (chopped *big.Int) {
+
+	// remove the negative and add it back when returning
+	if d.Sign() == -1 {
+		d = new(big.Int).Neg(d)
```

Comment:
Can we do this in-place?

Reply:
I originally had the logic flow to consider negatives throughout it (instead of performing only for the positive and then making it a negative again at the end) but that logic was actually a big headache, it's much cleaner to mutate and then unmutate, and I think its easier to understand given all the additionally if statements required if we don't do this. Had to do it this way with rat too. -> going to leave

Reply:
addressed by val